Northstar.Custom/mod/scripts/vscripts/ump/sh_ump_rpc.gnut

```diff
@@ -0,0 +1,458 @@
+
+		switch (type(data)) {
+		case "table":
+			table<var, var> dataTable
+			foreach (key, val in data)
+				dataTable[key] <- val
+			array<var> dataArray = UMP_TableToArray(dataTable)
+			foreach (val in dataArray)
+				if (!SendArgument(val))
+					return
+			break
+		case "array":
+			foreach (val in data)
+				if (!SendArgument(val))
+					return
+			break
+		default:
+			if (!SendArgument(data))
+				return
+			break
+		}
+	}
+	
+	RPCLog("Finishing state", stateString, stateId, "for player", player)
+
+	callFunc(player, "SCB_UMP_RPC_FinishState", stateId)
+}
+
+void function UMP_RPC_CallClientFunction(entity player, string name, ...)
+{
+	if (!IsValid(player) || IsDisconnected(player)) {
+		RPCLog("Attempted to send Client RPC", name, "to invalid/disconnected player", player)
+		return
+	}
+
+	array<var> args
+	for (int i = 0; i < vargc; i++)
+		args.append(vargv[i])
+
+	SendRPCData(Remote_CallFunction_NonReplay, player, "CLIENT", name, args)
+}
+
+void function UMP_RPC_CallUIFunction(entity player, string name, ...)
+{
+	if (!IsValid(player) || IsDisconnected(player)) {
+		RPCLog("Attempted to send UI RPC", name, "to invalid/disconnected player", player)
+		return
+	}
+
+	array<var> args
+	for (int i = 0; i < vargc; i++)
+		args.append(vargv[i])
+
+	SendRPCData(Remote_CallFunction_UI, player, "UI", name, args)
+}
+#endif
+
+string function EscapeString(string text)
+{
+	string result = ""
+	for (int i = 0; i < text.len(); i++) {
+		var char = text[i]
+
+		if (char <= 0x20)
+			result += format("%%%X", char)
+		else
+			result += format("%c", text[i])
+	}
+	return result
+}
+
+string function UnescapeString(string text)
+{
+	for (int c = 0x20; c > 0; c--) {
+		string escaped = format("%%%X", c)
+		string unescaped = format("%c", c)
+
+		while (text.find(escaped) != null)
+			text = StringReplace(text, escaped, unescaped)
+	}
+
+	return text
+}
+
+#if CLIENT || UI
+void function UMP_RPC_CallServerFunction(string name, ...)
+{
+#if UI
+	if (!IsConnected()) {
+		RPCLog("Attempted to send an RPC", name, "to server while disconnected")
+		return
+	}
+#endif
+	
+	if (!(name in file.functionToPoolId)) {
+		RPCLog("Attempted to send an unregistered RPC", name, "to server")
+		return
+	}
+
+	int functionId = file.functionToPoolId[name]
+	string args = ""
+
+	for (int i = 0; i < vargc; i++) {
+		var data = vargv[i]
+
+		RPCLog("Sending argument", data, "for function", name)
+
+		args += " "
+
+		switch (type(data)) {
+		case "string":
+			args += "s" + EscapeString(string(data))
+			break
+		case "bool":
+		case "float":
+		case "int":
+			args += type(data).slice(0, 1) + data.tostring()
+			break
+		case "null":
+			args += "n"
+			break
+		default:
+			RPCLog("RPC", name, "-", "Argument of type", type(data), "is unsupported.")
+			return
+			break
+		}
+	}
+
+	string cmd = "rpc " + functionId + args
+	RPCLog("Resulting command:", cmd)
+	
+#if UI
+	ClientCommand(cmd)
+#elseif CLIENT
+	GetLocalClientPlayer().ClientCommand(cmd)
```

Comment:
I have a hunch this may have a RCE bug since a string arg can have semicolons which don't get unescaped.